### PR TITLE
[controller] Implement rv-status-config-device-minor-controller

### PR DESCRIPTION
**Pull request by IvanOgurchenok**
## Description

Implemented `rv-status-config-device-minor-controller` for automatic assignment of unique `deviceMinor` to each `ReplicatedVolume` in the cluster.

The controller:
- Assigns minimum free `deviceMinor` value in range [0; 1048575] among all RVs
- Uses `Config` existence check instead of `deviceMinor` value check (since 0 is a valid value)
- Ensures uniqueness of `deviceMinor` for each RV
- Fills gaps in sequence (e.g., if 0, 2, 3 are taken, next will get 1)

Implementation includes:
- Controller with CREATE/UPDATE triggers for RVs without Config
- Reconciler with deviceMinor assignment logic and parallel processing handling via `PatchStatusWithConflictRetry`
- 8 unit tests using Ginkgo/Gomega covering happy path, edge cases, idempotency

## Why do we need it, and what problem does it solve?

The controller solves the task of automatic assignment of unique `deviceMinor` values for DRBD devices. The `deviceMinor` value is used for naming DRBD devices as `/dev/drbd<minor>` (e.g., `/dev/drbd0`, `/dev/drbd1`).

**Problem:** Without this controller, each `ReplicatedVolume` would need manual assignment of unique `deviceMinor`, which:
- Requires manual administrator intervention
- Creates risk of conflicts when assigning duplicate values
- Doesn't scale with large number of volumes

**Solution:** The controller automatically assigns minimum free `deviceMinor` value when RV is created, ensuring:
- Process automation
- Uniqueness guarantee among all RVs in cluster
- Gap filling for efficient use of value range

## What is the expected result?

After applying changes:

**MUST:**
- When creating new `ReplicatedVolume` without `status.config`, controller automatically assigns unique `deviceMinor` in `rv.status.config.deviceMinor`
- The `deviceMinor` value will be minimum free among all existing RVs (gaps are filled)
- Each RV gets unique `deviceMinor` in range [0; 1048575]
- Repeated reconcile doesn't change already assigned `deviceMinor` (idempotency)

**MUST NOT:**
- Two different RVs should not get the same `deviceMinor`
- Controller should not change `deviceMinor` if `Config` already exists (even if value = 0)

**Verification:**
1. Create several `ReplicatedVolume` resources
2. Verify that each got unique `deviceMinor` in `status.config.deviceMinor`
3. Ensure values fill gaps (if RV with deviceMinor=1 is deleted, next new RV will get 1, not next sequential)
4. Verify that repeated reconcile doesn't change already assigned `deviceMinor

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [ ] Changes were tested in the Kubernetes cluster manually.

### CODE REVIEW

File: docs/dev/controllers/rv_status_config_device_minor/SPEC_COMPLIANCE.md
Comment (asergunov) [inline docs/dev/controllers/rv_status_config_device_minor/SPEC_COMPLIANCE.md]: Это точно нужно?

File: images/controller/internal/controllers/registry.go [hunk 1]
```diff
@@ -19,15 +19,16 @@ package controllers
...
+	rvstatusconfigdeviceminor "github.com/deckhouse/sds-replicated-volume/images/controller/internal/controllers/rv_status_config_device_minor"
 	rvrdiskfulcount "github.com/deckhouse/sds-replicated-volume/images/controller/internal/controllers/rvr_diskful_count"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
 )
 
 var registry []func(mgr manager.Manager) error
 
 func init() {
 	registry = append(registry, rvrdiskfulcount.BuildController)
+	registry = append(registry, rvstatusconfigdeviceminor.BuildController)
```
Comment (asergunov) [inline images/controller/internal/controllers/registry.go]: Может стоило просто добавить свой контроллер в список не удаляя старый?

File: images/controller/internal/controllers/rv_status_config_device_minor/reconciler.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+			deviceMinor := item.Status.DRBD.Config.DeviceMinor
+			if deviceMinor >= minDeviceMinor && deviceMinor <= maxDeviceMinor {
+				usedDeviceMinors[deviceMinor] = true
+				// Check if this is the RV we're processing
+				if item.Name == rv.Name {
+					currentRVHasDeviceMinor = true
+				}
+			}
+		}
+	}
```
Comment (asergunov) [inline images/controller/internal/controllers/rv_status_config_device_minor/reconciler.go]: Эээ зачем листить все если нас интересует только с известным именем? Как будто Get достаточно?

А блин этот if все спутал. Давай просто чекнем rv?

File: images/controller/internal/controllers/rv_status_config_device_minor/reconciler.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+			currentRV.Status.DRBD.Config = &v1alpha3.DRBDResourceConfig{}
+		}
+
+		// Set the pre-calculated deviceMinor
+		currentRV.Status.DRBD.Config.DeviceMinor = availableDeviceMinor
+
+		return nil
+	}); err != nil {
+		return reconcile.Result{}, fmt.Errorf("updating RV status with deviceMinor: %w", err)
+	}
```
Comment (asergunov) [inline images/controller/internal/controllers/rv_status_config_device_minor/reconciler.go]: Вот это мне не очень понятно. На сколько это безопасно? Может делать просто cl.Patch или cl.Status().Update с выходом ошибкой для выхода на второй реконсайл?

File: images/controller/internal/controllers/rv_status_config_device_minor/reconciler.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+			deviceMinor := item.Status.DRBD.Config.DeviceMinor
+			if deviceMinor >= MinDeviceMinor && deviceMinor <= MaxDeviceMinor {
+				usedDeviceMinors[deviceMinor] = struct{}{}
+				// Check if this is the RV we're processing
+				if item.Name == rv.Name {
+					currentRVHasDeviceMinor = true
+				}
+			}
+		}
+	}
```
Comment (asergunov) [inline images/controller/internal/controllers/rv_status_config_device_minor/reconciler.go]: Два одновременных реконсайла вызовут рейс кондишн. Они могут завладеть одинаковым deviceMinor

Вижу несколько вариантов: 
- ограничить одним ондновременным реконсайлом
- Сделать список занятых под мутексом и выдавать оттуда. Однако его придется как-то сделать сначала. Можно по nil map например просекать этот кейс

File: images/controller/internal/controllers/rv_status_config_device_minor/reconciler_test.go [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+
+				By("Reconciling volume without deviceMinor to test gap filling")
+				result2, err2 := rec.Reconcile(ctx, RequestFor(rvGap4))
+				Expect(err2).NotTo(HaveOccurred(), "reconciliation should succeed")
+				Expect(result2).ToNot(Requeue(), "should not requeue after successful assignment")
+
+				By("Verifying gap filling: minimum free deviceMinor between 6 and 8")
+				updatedRV2 := &v1alpha3.ReplicatedVolume{}
+				Expect(cl.Get(ctx, client.ObjectKeyFromObject(rvGap4), updatedRV2)).To(Succeed(), "should get updated ReplicatedVolume")
+				Expect(updatedRV2).To(HaveField("Status.DRBD.Config.DeviceMinor", BeNumerically("==", 7)), "should assign deviceMinor 7 to fill gap between 6 and 8")
```
[the remark this answers is not in the document]
Comment (IvanOgurchenok) [reply images/controller/internal/controllers/rv_status_config_device_minor/reconciler_test.go]: Есть в спеке

[the remark this answers is not in the document]
Comment (IvanOgurchenok) [reply images/controller/internal/controllers/rv_status_config_device_minor/reconciler_test.go]: Есть в спеке:
```
Инициализировать свойство rv.status.config.deviceMinor минимальным свободным значением среди всех RV.
```

File: images/controller/internal/controllers/rv_status_config_device_minor/controller.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+func BuildController(mgr manager.Manager) error {
+	rec := NewReconciler(
+		mgr.GetClient(),
+		mgr.GetLogger().WithName(RVStatusConfigDeviceMinorControllerName).WithName("Reconciler"),
+	)
+
+	return builder.ControllerManagedBy(mgr).
+		Named(RVStatusConfigDeviceMinorControllerName).
+		For(&v1alpha3.ReplicatedVolume{}).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 1}).
```
Comment (asergunov) [inline images/controller/internal/controllers/rv_status_config_device_minor/controller.go]: Напиши, пожалуйста комментарий зачем эта строчка. Сейчас возникает вопрос при чтении

File: images/controller/internal/controllers/rv_status_config_device_minor/reconciler.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+	// usedDeviceMinors and assign a different value.
+	usedDeviceMinors := make(map[uint]struct{})
+	for _, item := range rvList.Items {
+		if item.Status != nil && item.Status.DRBD != nil && item.Status.DRBD.Config != nil {
+			deviceMinor := item.Status.DRBD.Config.DeviceMinor
+			if deviceMinor >= v1alpha3.RVMinDeviceMinor && deviceMinor <= v1alpha3.RVMaxDeviceMinor {
+				usedDeviceMinors[deviceMinor] = struct{}{}
+			}
+		}
+	}
```
Comment (asergunov) [inline images/controller/internal/controllers/rv_status_config_device_minor/reconciler.go]: А если на них есть deletion timestamp? Его надо обрабатывать или нет? Можно ли считать его порт свободным. Это бы выяснить и где-то зафиксировать. Сейчас при чтении возникает вопрос: это ошибка или так задумано. Вопроса возникать не должно

File: images/controller/internal/controllers/rv_status_config_device_minor/reconciler.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+	if changedRV.Status == nil {
+		changedRV.Status = &v1alpha3.ReplicatedVolumeStatus{}
+	}
+	if changedRV.Status.DRBD == nil {
+		changedRV.Status.DRBD = &v1alpha3.DRBDResource{}
+	}
+	if changedRV.Status.DRBD.Config == nil {
+		changedRV.Status.DRBD.Config = &v1alpha3.DRBDResourceConfig{}
+	}
+	changedRV.Status.DRBD.Config.DeviceMinor = *availableDeviceMinor
```
Comment (asergunov) [inline images/controller/internal/controllers/rv_status_config_device_minor/reconciler.go]: Мне от Саши прилетел поинт, что это уже проверено выше и можно не проверять больше.

File: images/controller/internal/controllers/rv_status_config_device_minor/controller.go [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+	)
+	// MaxConcurrentReconciles: 1
+	// prevents race conditions when assigning unique deviceMinor values
+	// to different ReplicatedVolume resources. Status not protected by optimistic locking,
+	// so we need to prevent parallel reconciles for avoiding duplicate assignments.
+
+	return builder.ControllerManagedBy(mgr).
+		Named(RVStatusConfigDeviceMinorControllerName).
+		For(&v1alpha3.ReplicatedVolume{}).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 1}).
```
Comment (asergunov) [inline images/controller/internal/controllers/rv_status_config_device_minor/controller.go]: Всё-равно далековато. Давай ближе

```suggestion
	
	return builder.ControllerManagedBy(mgr).
		Named(RVStatusConfigDeviceMinorControllerName).
		For(&v1alpha3.ReplicatedVolume{}).
		// prevents race conditions when assigning unique deviceMinor values
		// to different ReplicatedVolume resources. Status not protected by optimistic locking,
		// so we need to prevent parallel reconciles for avoiding duplicate assignments.
		WithOptions(controller.Options{MaxConcurrentReconciles: 1}).
```

File: images/controller/internal/controllers/rv_status_config_device_minor/reconciler.go (hunk 7 above)
Comment (AleksZimin) [reply images/controller/internal/controllers/rv_status_config_device_minor/reconciler.go]: Согласен с Ваней. Нужно все существующие учитывать, даже если они удаляются

File: images/controller/internal/controllers/rv_status_config_device_minor/reconciler.go (hunk 8 above)
Comment (AleksZimin) [reply images/controller/internal/controllers/rv_status_config_device_minor/reconciler.go]: Тут другая ситуация. Проверка выше есть, но она не завершает работу над ресурсом, у которого нет статуса. В эту часть может придти ресурс без статуса и это будет валидным кейсом

File: images/controller/internal/controllers/rv_status_config_device_minor/reconciler_test.go [hunk 10]
```diff
@@ -0,0 +1,426 @@
...
+			rv := &v1alpha3.ReplicatedVolume{
+				ObjectMeta: metav1.ObjectMeta{Name: "volume-1"},
+			}
+			Expect(cl.Create(ctx, rv)).To(Succeed(), "should create ReplicatedVolume")
+
+			By("Reconciling with Get interceptor that returns error")
+			internalServerError := errors.New("internal server error")
+			Expect(rec.Reconcile(ctx, RequestFor(rv))).Error().To(MatchError(internalServerError), "should return error when Get fails")
+		})
+	})
```
Comment (asergunov) [inline images/controller/internal/controllers/rv_status_config_device_minor/reconciler_test.go]:
```suggestion
	When("Get fails with non-NotFound error", func() {
		testError := errors.New("internal server error")
		BeforeEach(func() {
			clientBuilder = clientBuilder.WithInterceptorFuncs(
				InterceptGet(func(_ *v1alpha3.ReplicatedVolume) error {
					return testError
				}),
			)
		})

		It("should fail if getting ReplicatedVolume failed with non-NotFound error", func(ctx SpecContext) {
			By("Creating ReplicatedVolume")
			rv := &v1alpha3.ReplicatedVolume{
				ObjectMeta: metav1.ObjectMeta{Name: "volume-1"},
			}
			Expect(cl.Create(ctx, rv)).To(Succeed(), "should create ReplicatedVolume")

			By("Reconciling with Get interceptor that returns error")
			Expect(rec.Reconcile(ctx, RequestFor(rv))).Error().To(MatchError(testError), "should return error when Get fails")
		})
	})
```

File: images/controller/internal/controllers/rv_status_config_device_minor/reconciler_test.go [hunk 11]
```diff
@@ -0,0 +1,426 @@
...
+				}),
+			)
+		})
+
+		It("should fail if getting ReplicatedVolume failed with non-NotFound error", func(ctx SpecContext) {
+			By("Creating ReplicatedVolume")
+			rv := &v1alpha3.ReplicatedVolume{
+				ObjectMeta: metav1.ObjectMeta{Name: "volume-1"},
+			}
+			Expect(cl.Create(ctx, rv)).To(Succeed(), "should create ReplicatedVolume")
```
Comment (asergunov) [inline images/controller/internal/controllers/rv_status_config_device_minor/reconciler_test.go]: Значит можно тащить в `When("RV created"` а тут убирать. И ниже такой-же

**Review by asergunov — CHANGES_REQUESTED** [review]

**Review by asergunov — APPROVED** [review]
Было бы здорово использовать HaveField  чтобы убрать проверки на nil. Но это не блокер

File: images/controller/internal/controllers/rv_status_config_device_minor/reconciler.go [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+	if availableDeviceMinor == nil {
+		// All deviceMinors are used - this is extremely unlikely (1,048,576 volumes),
+		// but we should handle it gracefully
+		err := fmt.Errorf(
+			"no available deviceMinor for volume %s (all %d deviceMinors are used)",
+			rv.Name,
+			int(v1alpha3.RVMaxDeviceMinor-v1alpha3.RVMinDeviceMinor)+1,
+		)
+		log.Error(err, "no available deviceMinor for volume", "volume", rv.Name, "maxDeviceMinors", int(v1alpha3.RVMaxDeviceMinor-v1alpha3.RVMinDeviceMinor)+1)
```
[the remark this answers is not in the document]
Comment (IvanOgurchenok) [reply images/controller/internal/controllers/rv_status_config_device_minor/reconciler.go]: исправил

File: images/controller/internal/controllers/rv_status_config_device_minor/reconciler_test.go [hunk 13]
```diff
@@ -0,0 +1,604 @@
...
+					ObjectMeta: metav1.ObjectMeta{Name: "volume-dup-d1"},
+				}
+
+				Expect(cl.Create(ctx, rvA1)).To(Succeed(), "should create ReplicatedVolume A1")
+				Expect(cl.Create(ctx, rvA2)).To(Succeed(), "should create ReplicatedVolume A2")
+				Expect(cl.Create(ctx, rvB1)).To(Succeed(), "should create ReplicatedVolume B1")
+				Expect(cl.Create(ctx, rvB2)).To(Succeed(), "should create ReplicatedVolume B2")
+				Expect(cl.Create(ctx, rvB3)).To(Succeed(), "should create ReplicatedVolume B3")
+				Expect(cl.Create(ctx, rvC1)).To(Succeed(), "should create ReplicatedVolume C1")
+				Expect(cl.Create(ctx, rvD1)).To(Succeed(), "should create ReplicatedVolume D1")
```
[the remark this answers is not in the document]
Comment (IvanOgurchenok) [reply images/controller/internal/controllers/rv_status_config_device_minor/reconciler_test.go]: готова

File: images/controller/internal/controllers/rv_status_config_device_minor/reconciler_test.go [hunk 14]
```diff
@@ -0,0 +1,604 @@
...
+					g.Expect(updatedA1.Status.Errors.DuplicateDeviceMinor).ToNot(BeNil(), "A1 should have duplicate error")
+					g.Expect(updatedA1.Status.Errors.DuplicateDeviceMinor.Message).To(
+						And(
+							ContainSubstring("deviceMinor"),
+							ContainSubstring("0"),
+							ContainSubstring("is used by volumes:"),
+							ContainSubstring("volume-dup-a1"),
+							ContainSubstring("volume-dup-a2"),
+						),
+					)
```
[the remark this answers is not in the document]
Comment (IvanOgurchenok) [reply images/controller/internal/controllers/rv_status_config_device_minor/reconciler_test.go]: готова
